### PR TITLE
MSP over TBS Crossfire

### PR DESCRIPTION
**Pull request by teckel12**
#3886 Tests good so far.  Doing more testing today and should be ready for merging.

### CODE REVIEW

**Comment by teckel12** [comment]
Only "failed" on the OMNIBUS target due to memory usage.  But it compiles fine for me (minus the latest development merges that my fork doesn't have).

With my local build, I show 40,940 bytes used (20 bytes free).

**Comment by teckel12** [comment]
I did the OMNIBUS build without `crsfFrameDeviceInfo` from the latest commit and it still uses 40,932 bytes (only saves 8 bytes).  So, it seems to add MSP over Crossfire something will have to give (maybe disabling default features on the F3?)

**Comment by stronnag** [comment]
Perhaps you just don't enable this on F3, it is slightly specialised after all.

**Comment by shellixyz** [comment]
It compiles fine it is just an issue with the version of the compiler used by Travis. Maybe it needs to be updated. I tried with 7.1.0 and it fits.

File: src/main/fc/config.h [hunk 1]
```diff
@@ -79,6 +79,7 @@ typedef struct systemConfig_s {
     uint8_t throttle_tilt_compensation_strength;      // the correction that will be applied at throttle_correction_angle.
     inputFilteringMode_e pwmRxInputFilteringMode;
     char name[MAX_NAME_LENGTH + 1];
+   	char boardIdentifier[sizeof(TARGET_BOARD_IDENTIFIER) + 1];
```
Comment (shellixyz) [inline src/main/fc/config.h]: Please fix the indentation

**Comment by teckel12** [comment]
@stronnag If needed, I'm not against disabling this on the F3 processor.  But, we probably shouldn't disable all new features going forward on the F3.  Instead, we should probably look at everything and decide what's the least used that uses the most memory and start with disabling that.

**Comment by shellixyz** [comment]
@teckel12 Like the OLED screen stuff

**Comment by teckel12** [comment]
@shellixyz Exactly what I was thinking.

**Comment by tgreer** [comment]
Is the benefit of MSP over crossfire to relay it via bluetooth to a device? Or can you do stuff from opentx via MSP?

**Comment by wx4cb** [comment]
@tgreer i use the bluetooth in my full module for that reason, i can config the FC wirelessly and also use something like mobile flight to get telemetry and a cute map (when it works :D)

**Comment by teckel12** [comment]
@tgreer For example, with MSP over Crossfire you can set VTx band/channel/power from your transmitter.

**Comment by teckel12** [comment]
I tested this today again and was able to use the Betaflight Lua Script to set VTx over Crossfire and reboot the FC. 

Appears good to merge from multiple testers. We can always enhance it later but for now it works well and is beneficial.

I'll open a new issue for making OLED disabled by default on the F3 as that seems like a good place to save memory as it's seldom used/not very useful.

**Comment by tgreer** [comment]
@teckel12 nice work :)

**Comment by fiam** [comment]
Merging #4090 should make enough room for this on F3

**Comment by teckel12** [comment]
@fiam :thumbsup:

We should also future-plan by considering features to disable on the F3 by default. OLED, sonar and optical flow are the 3 current contenders.

**Comment by fiam** [comment]
Both rangefinder and opflow are already disabled on F3. OLED would save ~3.2K flash and ~0.8K RAM.

**Comment by teckel12** [comment]
@fiam Now that some space is saved on the F3, this should merge/build. I've tested in with Betaflight Lua Telemetry and it works fine, using it now for VTx config.

File: src/main/fc/config.h [hunk 2]
```diff
@@ -79,6 +79,7 @@ typedef struct systemConfig_s {
     uint8_t throttle_tilt_compensation_strength;      // the correction that will be applied at throttle_correction_angle.
     inputFilteringMode_e pwmRxInputFilteringMode;
     char name[MAX_NAME_LENGTH + 1];
+    char boardIdentifier[sizeof(TARGET_BOARD_IDENTIFIER) + 1];
```
Comment (fiam) [inline src/main/fc/config.h]: Don't add this here just to output it in `crsfFrameDeviceInfo()`, use `TARGET_BOARD_IDENTIFIER` in that function directly.

File: src/main/rx/crsf.h [hunk 3]
```diff
@@ -42,27 +51,61 @@ enum {
...
     CRSF_ADDRESS_RESERVED2 = 0xCA,
     CRSF_ADDRESS_RACE_TAG = 0xCC,
     CRSF_ADDRESS_RADIO_TRANSMITTER = 0xEA,
     CRSF_ADDRESS_CRSF_RECEIVER = 0xEC,
     CRSF_ADDRESS_CRSF_TRANSMITTER = 0xEE
-};
+} crsfAddress_e;
+
+// #define CRSF_PAYLOAD_SIZE_MAX   62
+// #define CRSF_FRAME_SIZE_MAX     (CRSF_PAYLOAD_SIZE_MAX + 4)
```
Comment (fiam) [inline src/main/rx/crsf.h]: Delete these

File: src/main/rx/crsf.h [hunk 4]
```diff
@@ -42,27 +51,61 @@ enum {
...
+    CRSF_FRAMETYPE_COMMAND = 0x32,
+    // MSP commands
+    CRSF_FRAMETYPE_MSP_REQ = 0x7A,   // response request using msp sequence as command
+    CRSF_FRAMETYPE_MSP_RESP = 0x7B,  // reply with 58 byte chunked binary
+    CRSF_FRAMETYPE_MSP_WRITE = 0x7C,  // write with 8 byte chunked binary (OpenTX outbound telemetry buffer limit)
+    CRSF_FRAMETYPE_DISPLAYPORT_CMD = 0x7D, // displayport control command
+} crsfFrameType_e;
 
-typedef struct crsfFrameDef_s {
+ typedef struct crsfFrameDef_s {
```
Comment (fiam) [inline src/main/rx/crsf.h]: Delete the space at the start of the line

**Comment by teckel12** [comment]
@fiam Issues addressed.

**Comment by teckel12** [comment]
@shellixyz Still a chance of getting this in 2.1 now that the F3 memory usage has been reduced?

**Comment by shellixyz** [comment]
May be possible. I'll ask the others. The SPRACINGF3NEO still fails to build though (RAM overflowed by 8 bytes) we would need to disable more stuff on F3 for it to fit.

**Comment by teckel12** [comment]
@shellixyz I thought we disabled some seldom used features on the F3 that used quite a bit of RAM?

I'll look at disabling something on F3 or maybe even MSP over telemetry. I could also save 8 bytes by removing one of my commits too.

**Comment by wx4cb** [comment]
What about things that arent used telemetry wise. Things like ltm hott and so forth.

**Comment by stronnag** [comment]
> What about things that arent used telemetry wise. Things like ltm hott and so forth.

ltm is a widely used telemetry protocol.

**Comment by teckel12** [comment]
@shellixyz `make TARGET=SPRACINGF3NEO` doesn't work locally?  I get the error:

```
Target 'SPRACINGF3NEO' is not valid, must be one of ...
```

**Comment by teckel12** [comment]
@shellixyz So looking at other SP Racing F3 targets they're nowhere close to running out of memory.  So, the SP Racing F3 Neo is just really close to running out of RAM, even without this PR.

There's already been some tweaking of the target to get it to build.  Certain features have already been disabled for this target only.  So, I went with the already established method and disabled the pitot sensor on the SP Racing F3 NEO (figuring it would be rare for that combination)

**Comment by teckel12** [comment]
@shellixyz @fiam It builds fine now, and I believe disabling the pitot sensor for this FC is the best thing to do (and it seems @digitalentity has gone this route as well: https://github.com/iNavFlight/inav/commit/ff0746d5e53cd49d526b8f5ea67cf20530e6129e, https://github.com/iNavFlight/inav/commit/0de7f538fb37924bbcdc253b19a98ae652d026d6).

**Comment by teckel12** [comment]
@shellixyz @fiam Could we get this merged in 2.1.0?  It's been heavily tested by both myself and a few others.  It also makes Crossfire telemetry updates **MUCH** faster (in-line with Betaflight).  I've got a few people depending on it (as well as myself) 😀 Thanks!

**Review by digitalentity — APPROVED** [review]
LGTM

**Comment by ZinhoRacing** [comment]
look like it is not in the 2.1.0 release :disappointed:

**Comment by teckel12** [comment]
@ZinhoRacing It was merged to 2.1.0 RC3, so it should be in 2.1.0 release as well.

**Comment by teckel12** [comment]
@ZinhoRacing It's in the master branch.  I haven't tested the 2.1.0 release yet, but I did test it in 2.1.0 RC3 and it worked.

**Comment by ZinhoRacing** [comment]
@teckel12 my mistake, it is working fine. 😅 
Thanks for your time to made the PR and tests !

**Comment by putimir** [comment]
@teckel12 : Hoorah! Thank you very much! 

BTW - since you are the "master" of all these LUA/MSP scripts, and you now own a horus radio - are you workin on some horus widget to to take advantage of the screen/and widget functionality and port these scripts to widgets?

Again, good job! You are earning some serious karma points hehe

**Comment by teckel12** [comment]
@putimir There is a bit of a roadblock with widgets on the Horus. OpenTX doesn't allow a widget to receive any key inputs, so a widget can't be controlled by the user. This is an OpenTX limitation that many people wish would be fixed. But, an OpenTX issue, not INAV so this is the wrong place for this.

**Pull request merged**